command/agent/consul/client.go

```diff
@@ -0,0 +1,636 @@
+				}
+				//TODO Log? and jitter/backoff
+				timer.Reset(c.retryInterval)
+			} else {
+				if !lastOk {
+					c.logger.Printf("[INFO] consul: successfully updated services in Consul")
+					lastOk = true
+				}
+			}
+		case <-c.shutdownCh:
+			return
+		}
+	}
+}
+
+// forceSync asynchronously causes a sync to happen. Any operations enqueued
+// prior to calling forceSync will be synced.
+func (c *ServiceClient) forceSync() {
+	select {
+	case c.syncCh <- mark:
+	default:
+	}
+}
+
+// sync enqueued operations.
+func (c *ServiceClient) sync() error {
+	// Shallow copy and reset the pending operations fields
+	c.regLock.Lock()
+	regServices := make(map[string]*api.AgentServiceRegistration, len(c.regServices))
+	for k, v := range c.regServices {
+		regServices[k] = v
+	}
+	c.regServices = map[string]*api.AgentServiceRegistration{}
+
+	regChecks := make(map[string]*api.AgentCheckRegistration, len(c.regChecks))
+	for k, v := range c.regChecks {
+		regChecks[k] = v
+	}
+	c.regChecks = map[string]*api.AgentCheckRegistration{}
+
+	regScripts := make(map[string]*scriptCheck, len(c.regScripts))
+	for k, v := range c.regScripts {
+		regScripts[k] = v
+	}
+	c.regScripts = map[string]*scriptCheck{}
+
+	deregServices := make(map[string]struct{}, len(c.deregServices))
+	for k := range c.deregServices {
+		deregServices[k] = mark
+	}
+	c.deregServices = map[string]struct{}{}
+
+	deregChecks := make(map[string]struct{}, len(c.deregChecks))
+	for k := range c.deregChecks {
+		deregChecks[k] = mark
+	}
+	c.deregChecks = map[string]struct{}{}
+	c.regLock.Unlock()
+
+	var err error
+
+	regServiceN, regCheckN, deregServiceN, deregCheckN := len(regServices), len(regChecks), len(deregServices), len(deregChecks)
+
+	// Register Services
+	for id, service := range regServices {
+		if err = c.client.ServiceRegister(service); err != nil {
+			goto ERROR
+		}
+		delete(regServices, id)
+	}
+
+	// Register Checks
+	for id, check := range regChecks {
+		if err = c.client.CheckRegister(check); err != nil {
+			goto ERROR
+		}
+		delete(regChecks, id)
+
+		// Run the script for this check if one exists
+		if script, ok := regScripts[id]; ok {
+			// This check is a script check; run it
+			c.runningScripts[id] = script.run()
+		}
+	}
+
+	// Deregister Checks
+	for id := range deregChecks {
+		if h, ok := c.runningScripts[id]; ok {
+			// This check is a script check; stop it
+			h.cancel()
+			delete(c.runningScripts, id)
+		}
+
+		if err = c.client.CheckDeregister(id); err != nil {
+			goto ERROR
+		}
+		delete(deregChecks, id)
+	}
+
+	// Deregister Services
+	for id := range deregServices {
+		if err = c.client.ServiceDeregister(id); err != nil {
+			goto ERROR
+		}
+		delete(deregServices, id)
+	}
+
+	c.logger.Printf("[DEBUG] consul: registered %d services / %d checks; deregisterd %d services / %d checks", regServiceN, regCheckN, deregServiceN, deregCheckN)
+	return nil
+
+	//TODO Labels and gotos are nasty; move to a function?
```

Comment:
add helper functions to get a snapshot of the maps and another function to merge outstanding functions back in